### ISSUE TITLE
grating monochromator with  variable period 

### ISSUE THREAD

**Issue by yangfg-bsrf** [kklmn/xrt#73]
Hi, recently I am trying to simulate the grating monochromator with variable period.  The code is modified referring to the Softi_CXIw2D.py in the XRT example. However, I found the result changes with the run, sometimes, no beam output by M3, with following information.

> "
> Warning: the bandwidth seems too big. Specify it by giving eMin and eMax in the constructor.
> Alarm! M3 absorbes 100.00% of rays at 10% alarm level!
> The ray tracing with 1 iteration took 6.3 s
> finished
> "

and it works somtime as follow.
![BC-2D-1-rays-1000eV-0emit-0enSpread-monoE-06i-ExpFocus-Is00](https://user-images.githubusercontent.com/41823786/110323278-5b738080-804f-11eb-8562-4a3ec07cd8af.png)
 would you please help me check the reason for it? Thanks.

the code files(the .txt should be changed to .py in order to run them):
[BC_V4py.txt](https://github.com/kklmn/xrt/files/6101378/BC_V4py.txt)

[Distorted_Element.txt](https://github.com/kklmn/xrt/files/6101376/Distorted_Element.txt)

**Comment by yangfg-bsrf** [comment]
**- I have checked the layout of the beamline, and put a screen on the position of  M3， it is found that the  plot is OK if not reflected by the M3. as shown below**

> M3-no-reflection
> <img src= https://user-images.githubusercontent.com/41823786/110576347-971b6100-819b-11eb-9378-0f265e0bf1e4.png width="512"> )
> [M3-local]
> <img src= https://user-images.githubusercontent.com/41823786/110576352-984c8e00-819b-11eb-9a20-455b8167684a.png  width="512"> )
> sample
> <img src= https://user-images.githubusercontent.com/41823786/110576353-98e52480-819b-11eb-8ec9-95380b39da70.png width="512"> 
> 

**but, if the run is not OK, the output is :**

> M3-no-reflection
> <img src= https://user-images.githubusercontent.com/41823786/110576854-73a4e600-819c-11eb-9580-8d4702981055.png width="512"> )
> [M3-local]
> <img src= https://user-images.githubusercontent.com/41823786/110576878-828b9880-819c-11eb-9b2f-2df0cecc6812.png  width="512"> )
> sample
> <img src= https://user-images.githubusercontent.com/41823786/110576911-96cf9580-819c-11eb-939f-b9c60103b89c.png width="512"> 

so, the alignment of the mirror is found no problem.

**Comment by yangfg-bsrf** [comment]
and when I checked the the direction of the output beam of PGM, it is found that the direction b component of some rays are negative and some are positive, whereas all the 'state' =1 . I am not sure it is the problem.

**Comment by kklmn** [comment]
To see what is wrong with your alignment, remove limPhysX and limPhysY from your optical elements. Also remove limits from the plots. 

You use a negative grating order. My definition of the sign depends on cff, see `align_grating` where m is + or - order. If your definition is different, change the grating alignment accordingly.

Grating alignment depends on line density. You use VLS, where the actual line density is not the constant `rho` used in `align_grating`. You should either calculate line density for the beam center or iteratively find its value that maintains your fixed exit.

**Comment by yangfg-bsrf** [comment]
For the grating parameter, the `gratingDensity = ["y", rho, 1, 0.078/rho/2, 2.7e-6/rho/3]` agree with the theory. 
Possiblely, the order m is negative because I did not use the following definition:
```
class Grating(roe.OE):
    def local_g(self, x, y, rho=rho):
        return 0, -rho, 0  # constant line spacing along y
```
I have check the alignment and remove the  `limPhysX` and `limPhysY` from the optics. But the problem is always over there.
I will check it again and again to make them physically correct, not accidentally (￣▽￣)".

### what I am really confused is what will cause the different results for different run even though nothing changes.  If the alignment is wrong, there should be no light in sample position all the time.

**Comment by kklmn** [comment]
> Possiblely, the order m is negative because I did not use the following definition...

Yes, indeed, if your rho is positive (`gratingDensity[1]>0`), the order should be negative and vice versa.

**Comment by yangfg-bsrf** [comment]
thanks for your suggestion. You re right this is my mistake, because I have little experience in grating design. In the past few year,  I focus on the hard x-ray beamlines. So this should be the first lesson I learn.  Thanks!
> But in fact I doubt you need `roe.VLSGrating` since you don't specify its essential parameters: _aspect_ and _depth_ that define intersections with the groove walls. You can use `roe.OE` instead. With `roe.OE`, you _can_ use `gratingDensity` outside of the constructor, although I don't see why you would need this: the coefficients there do not depend on energy or anything else, they are just constants.

**Comment by yangfg-bsrf** [comment]
I have tested it by specifying `gratingDensity` as a constructor keyword. but it didnot work. I am not sure if it is my fault.
But, following the example given, I change the type of the grating to the `BlazedGrating` type by setting the `blaze = 1e-9`. the result is shown below, it seems right.
![image](https://user-images.githubusercontent.com/41823786/110642709-3b2df800-81ee-11eb-9d7d-330a7cf00338.png)

> If you want to use `roe.VLSGrating`, you should specify `gratingDensity` as a constructor keyword, as the actual groove positions are calculated there, in the constructor. If you specify `gratingDensity` outside of the constructor, the grooves are specified for a set of default parameters, and the reflection behaviour is unpredicted.
> 
> In general, when you use an object field directly (like you do in `beamLine.pg.gratingDensity = ["y", rho, 1, 0.078/rho/2, 2.7e-6/rho/3]`), you cannot know the result of it, unless you see an example. Documented is only its use in the class instantiation.

**Comment by yangfg-bsrf** [comment]
@kklmn I have another issue is about the wave propagation, have you got any wave propagation example or test result for PGM with variable density? I have tested the program following the example. but is does not work.

`    #M1
    beamM1global, beamM1local = beamLine.m1.reflect(beamSource)
    #M2
    waveOnm2 = beamLine.m2.prepare_wave(beamLine.m1, nrays)
    beamTom2 = rw.diffract(beamM1local, waveOnm2)      
    beamM2global, beamM2local = beamLine.m2.reflect(beamTom2, noIntersectionSearch=True)
    # beamM2global, beamM2local = beamLine.m2.reflect(beamM1global)
    
    #PG
    beamPGglobal, beamPGlocal = beamLine.pg.reflect(beamM2global)
    
    #M3
    waveOnm3 = beamLine.m3.prepare_wave(beamLine.pg, nrays)
    beamTom3 = rw.diffract(beamPGlocal, waveOnm3) 
    beamM3global, beamM3local = beamLine.m3.reflect(beamTom3)
    # beamM3global, beamM3local = beamLine.m3.reflect(beamPGglobal)
    
    beamPGlocal.area = 0
    beamPGlocal.areaFraction = beamLine.pg.areaFraction
    
    waveOnExitSlit = beamLine.exitSlit.prepare_wave(beamLine.m3, nrays)
    rw.diffract(beamM3local, waveOnExitSlit)
    beamExitSlit = waveOnExitSlit
    
    for waveOnSample in waveOnSamples:
        rw.diffract(beamM3local, waveOnSample)`
ExitSlit
![BC-2D-2-hybr-1000eV-0emit-0enSpread-monoE-03-ExitSlit](https://user-images.githubusercontent.com/41823786/111065348-230eef00-84f4-11eb-8d40-0a7dd4522a49.png)
M3local
![BC-2D-2-hybr-1000eV-0emit-0enSpread-monoE-02-M3local](https://user-images.githubusercontent.com/41823786/111065336-19858700-84f4-11eb-9b16-a96b7b05159a.png)
pglocal
![BC-2D-2-hybr-1000eV-0emit-0enSpread-monoE-01-pglocal](https://user-images.githubusercontent.com/41823786/111065326-0d012e80-84f4-11eb-838e-10e74787595c.png)

**Comment by kklmn** [comment]
What exactly does not work? Is your concern about some random featureless illumination of optical elements that receive diffraction? Then you need more field samples, and if you cannot afford them then you should consider 1D cuts, as in the 
[SoftiMAX example](https://xrt.readthedocs.io/gallery3.html#cxi-branch).

**Comment by yangfg-bsrf** [comment]
sorry for my unclear question, in the previous simulation, the spot in the exitslit has unexpected non-uniformity. But when I recheck the code following your advice, the result is good if I use the hybrid method( only using wave propagation after M3). and for the wave mode,  some issues exist and should be checked in the next step.

**Comment by yangfg-bsrf** [comment]
Hi, I think I get the basic problem for the code. In my previous code, the hybrid mode is used, the geometric ray tracing is used for M2->pg because the distance between M2 and pg is so small. And the wave propagation is used for m1->m2, pg->m3 and m3->exitslit. but the code can not run for pg. I wonder if this kind of hybrid mode can work in XRT in principle?

**Comment by kklmn** [comment]
Yes, an example of a grating in waves is given [here](https://xrt.readthedocs.io/gallery3.html#diffraction-from-grating). If you want a grating in wave propagation, it must have a physical periodic depth profile, not just a grating vector, as it would be sufficient for rays.
As to doing diffraction from m1 to m2, I don't see any value in it. All wave effects are only visible from the last optical elements. This is shown in the example of SoftiMAX.

**Comment by yangfg-bsrf** [comment]
The following picture is the optical layout of the beamline, the distance from m1 to m2 is about 7m. I am trying to test the diffraction effect for this case.
![image](https://user-images.githubusercontent.com/41823786/111237546-42ab3200-8630-11eb-88ec-7cbe1cd94be5.png)

**Comment by yangfg-bsrf** [comment]
I have tested the code in the example, this kind of hybrid mode can work. So my problem is the grating you mentioned. Thanks!

> Hi, I think I get the basic problem for the code. In my previous code, the hybrid mode is used, the geometric ray tracing is used for M2->pg because the distance between M2 and pg is so small, where the wave propagation can not work. The wave propagation is used for m1->m2, pg->m3 and m3->exitslit. but the code can not run for pg. I wonder if this kind of hybrid mode can work in XRT in principle?

> Yes, an example of a grating in waves is given [here](https://xrt.readthedocs.io/gallery3.html#diffraction-from-grating). If you want a grating in wave propagation, it must have a physical periodic depth profile, not just a grating vector, as it would be sufficient for rays.
> As to doing diffraction from m1 to m2, I don't see any value in it. All wave effects are only visible from the last optical elements. This is shown in the example of SoftiMAX.
